### ISSUE TITLE
Why is pyannote not using my GPU ro CPU? So slow too.

### ISSUE THREAD

**Issue by CrackerHax** [pyannote/pyannote-audio#1702]
### Tested versions

latest version 3.1

### System information

windows 11, amd 5950x cpu, ubuntu 20.04 python 3.9 latest pyannote 3.1

### Issue description

CPU at 10%, both Nvidia rtx3080s at 0%, I only see the model taking up GPU memory and it is very small. Using the sample code provided in the README

### Minimal reproduction example (MRE)

Use the sample speaker diarization README code

```
diarization = pyannote.audio.Pipeline.from_pretrained(
    "pyannote/speaker-diarization-3.1",
    use_auth_token="<my token>").to(torch.device("cuda"))

diarization_output = diarization(filename)
```

![image](https://github.com/pyannote/pyannote-audio/assets/6037535/e3a35e5c-f131-4574-a498-06fba65e91c0)
![image](https://github.com/pyannote/pyannote-audio/assets/6037535/11c9eede-a2ed-4d8b-bff4-7006faee596b)

**Comment by freshpearYoon** [comment]
Hi, I am facing the same issue. Did you solve the problem?

**Comment by CrackerHax** [comment]
> Hi, I am facing the same issue. Did you solve the problem?

No I didn't. I just ended up letting it take forever.

**Comment by amas0** [comment]
Just wanted to chime in and say that I'm seeing similar-ish issues. I do see utilization on my GPU, so might not be the same thing, but general performance issues. I may have localized the issue to passing an audio file path to the pipeline directly. 

TL;DR -- try preprocessing your audio into a waveform and running it that way. I dropped my processing from 50 seconds down to 12 seconds by doing so.

I had a 3 minute clip I used as a test here. Passing it into the pipeline as a path, e.g.

```python

pipline = Pipeline.from_pretrained(...)
pipline('audio.mp3')
```

took about 50 seconds overall. Doing some profiling, I found that the code was spending a ton of time in the function `pyannote.audio.core.io.Audio.crop`. 

Specifically this snippet here:

https://github.com/pyannote/pyannote-audio/blob/6a972c0c4e95de04637d7221208736c64c8b972a/pyannote/audio/core/io.py#L341-L354

where passing an audio file directly follows the `else:` block and spends a lot of time seemingly doing file I/O by loading the file for `get_torchaudio_info`. The docstring of that function claims that it should cache the output `torchaudio.info` but I couldn't quite grok where it was cacheing it. Manually implementing a cache there lowered by time to run down to 33 secs or so. So a big chunk, but not everything. I poked a bit more to see what was going on and it still seemed to have all the performance issues in that one block for the audio file processing.

Insteading of running it completely to ground since it seemed to mostly be a problem with audio file processing, I tried preprocessing via a torchaudio load as recommended on the HF page:

```py
waveform, sample_rate = torchaudio.load("audio.wav")
diarization = pipeline({"waveform": waveform, "sample_rate": sample_rate})
```

This resulted in my 6 min clip being diarized in 12 seconds and I seed solid GPU utilization along the way.

**Comment by melMass** [comment]
Probably not directly related but the way pyannote is currently packaged overwrites your environment torch and installs the CPU version instead... I spent way too long tracking it down to this project so that might be it...

You can quickly check with: `python -c "import torch;print(torch.cuda.is_available())" `

**Comment by replic1111111** [comment]
you need to install the cuda libraries and reinstall pyannote. There are 2 seperate cuda related nvidia libraries that need to be installed before it properly detects cuda and uses gpu.

**Comment by cjbehm** [comment]
It's too bad this has been marked closed; I just stumbled onto this exact performance bottleneck.

At least this thread exists; I also got a 4x improvement in throughput

**Comment by caixxiong** [comment]
Me too, it use a lots of cpu, but gpu utilization rate is very low. I use bash script start 15 pyannnote-python processes. 

<img width="2658" height="620" alt="Image" src="https://github.com/user-attachments/assets/1a3a16d0-4b7b-4be1-9f88-cd0b57a61a4e" />

<img width="1840" height="932" alt="Image" src="https://github.com/user-attachments/assets/7302c04d-7254-46c5-b03d-d2db619bba67" />


Here is my python code:
```python

def run(args):
    rank = args.rank
    device = torch.device(f'cuda:{rank}')
    torch.cuda.set_device(device)

    print(f'[INFO] Worker with {rank=}, model={args.model_name}\n', flush=True)

    input_files = load_meta(args.input_list)
    if len(input_files) == 0:
        print(f'[WARN] total audio files is 0 for {args.input_file}, will do nothing.')
        return

    print(f'[INFO] Loading pyannote diar model: {args.model_name} ..')
    pipeline = Pipeline.from_pretrained(args.model_name)
    pipeline.to(device)

    res_lines = []
    for i, audio_file in tqdm(enumerate(input_files), total=len(input_files), desc=f'Pyannote diarization({args.model_name}) ..'):
        try:
            waveform, sample_rate = torchaudio.load(audio_file)
            with torch.inference_mode():
                res = pipeline({'waveform': waveform, 'sample_rate': sample_rate})
            segs = [f'{turn.start:.2f} {turn.end:.2f} {spk.lower()}' for turn, spk in res.speaker_diarization]
            n_spks = len({s.split()[-1] for s in segs})
            line = f'{audio_file}|{n_spks}|{segs}\n'
            res_lines.append(line)
        except Exception as e:
            print(f'ERROR: failed by {e}|{audio_file}')

    header = f'#wav_file|num spks| spk time_stamps(from model: [{args.model_name}])\n'
    res_lines = sorted(res_lines, key=lambda x: int(x.strip().split('|')[1]))
    res_lines = [header] + res_lines
    with open(args.output_file, 'w') as fw:
        fw.writelines(res_lines)

    print(f'Rank={rank} process finished, output saved at {args.output_file}')

```
